providers/storj/storj.go

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package storj
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/objstore"
+	"gopkg.in/yaml.v2"
+	"storj.io/common/fpath"
+	"storj.io/uplink"
+)
+
+// Config stores the configuration for the storj bucket.
+type Config struct {
+	AccessGrant string `yaml:"access"`
+	Bucket      string `yaml:"bucket"`
+}
+
+// Bucket implements the store.Bucket interface against storj API.
+type Bucket struct {
+	logger  log.Logger
+	project *uplink.Project
+	bucket  *uplink.Bucket
+}
+
+// NewBucket returns a new Bucket using the provided storj config values.
+func NewBucket(logger log.Logger, config []byte, component string) (*Bucket, error) {
+	var conf Config
+
+	err := yaml.Unmarshal(config, &conf)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewBucketWithConfig(logger, conf, component)
+}
+
+// NewBucketWithConfig returns a new Bucket using the provided storj config struct.
+func NewBucketWithConfig(logger log.Logger, config Config, component string) (*Bucket, error) {
+	var instance Bucket
+
+	ctx := fpath.WithTempData(context.TODO(), "", true)
+
+	uplConf := &uplink.Config{
+		UserAgent: fmt.Sprintf("thanos-%s", component),
+	}
+
+	parsedAccess, err := uplink.ParseAccess(config.AccessGrant)
+	if err != nil {
+		return nil, err
+	}
+
+	instance.project, err = uplConf.OpenProject(ctx, parsedAccess)
+	if err != nil {
+		return nil, err
+	}
+
+	instance.bucket, err = instance.project.EnsureBucket(ctx, config.Bucket)
+	if err != nil {
+		//Ignoring the error to return the one that occurred first, while still trying to clean up.
+		_ = instance.project.Close()
+		return nil, err
+	}
+
+	instance.logger = logger
+
+	return &instance, nil
+}
+
+// Name returns the name of the bucket.
+func (b *Bucket) Name() string {
+	return b.bucket.Name
+}
+
+// Close closes the Bucket.
+func (b *Bucket) Close() error {
+	return b.project.Close()
+}
+
+// Iter calls f for each entry in the given directory (not recursive). The argument to f is the full
+// object name including the prefix of the inspected directory.
+func (b *Bucket) Iter(ctx context.Context, dir string, f func(string) error, options ...objstore.IterOption) error {
+	_ = level.Debug(b.logger).Log("Listing %s from Storj Bucket", dir)
```

Comment:
Maybe we can remove these `level.Debug(...).Log()` calls? They would be quite spammy.